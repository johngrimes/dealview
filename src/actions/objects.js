// @flow

import type { ObjectMap, ObjectWithId } from 'types/commonTypes'

type PutObjectRequestAction = {
  +type: string,
  +object: Object,
}
type PutObjectSuccessAction = {
  +type: string,
  +object: ObjectWithId,
}
type PutObjectFailureAction = {
  +type: string,
  +error?: string,
}

type DeleteObjectRequestAction = {
  +type: string,
  +id: string,
}
type DeleteObjectSuccessAction = {
  +type: string,
  +id: string,
}
type DeleteObjectFailureAction = {
  +type: string,
  +error?: string,
}

type LoadObjectsRequestAction = {
  +type: string,
}
type LoadObjectsSuccessAction = {
  +type: string,
  +objects: ObjectMap,
}
type LoadObjectsFailureAction = {
  +type: string,
  +error?: string
}

export type ObjectAction = PutObjectRequestAction
                         | PutObjectSuccessAction
                         | PutObjectFailureAction
                         | DeleteObjectRequestAction
                         | DeleteObjectRequestAction
                         | DeleteObjectFailureAction
                         | LoadObjectsRequestAction
                         | LoadObjectsSuccessAction
                         | LoadObjectsFailureAction

export type ObjectActions = { +[string]: (any) => ObjectAction }

export const createObjectActions = (name: string, pluralName: string,
                                    nameTitled: string, pluralNameTitled: string): ObjectActions => {
  return {
    [`put${nameTitled}Request`]: (object: Object): PutObjectRequestAction => ({
      type: `PUT_${name}_REQUEST`,
      object,
    }),
    [`put${nameTitled}Success`]: (object: ObjectWithId): PutObjectSuccessAction => ({
      type: `PUT_${name}_SUCCESS`,
      object,
    }),
    [`put${nameTitled}Failure`]: (error?: string): PutObjectFailureAction => ({
      type: `PUT_${name}_FAILURE`,
      error,
    }),
    [`delete${nameTitled}Request`]: (id: string): DeleteObjectRequestAction => ({
      type: `DELETE_${name}_REQUEST`,
      id,
    }),
    [`delete${nameTitled}Success`]: (id: string): DeleteObjectSuccessAction => ({
      type: `DELETE_${name}_SUCCESS`,
      id,
    }),
    [`delete${nameTitled}Failure`]: (error?: string): DeleteObjectFailureAction => ({
      type: `DELETE_${name}_FAILURE`,
      error,
    }),
    [`load${pluralNameTitled}Request`]: (): LoadObjectsRequestAction => ({
      type: `LOAD_${pluralName}_REQUEST`,
    }),
    [`load${pluralNameTitled}Success`]: (objects: ObjectMap): LoadObjectsSuccessAction => ({
      type: `LOAD_${pluralName}_SUCCESS`,
      objects,
    }),
    [`load${pluralNameTitled}Failure`]: (error?: string): LoadObjectsFailureAction => ({
      type: `LOAD_${pluralName}_FAILURE`,
      error,
    }),
  }
}
