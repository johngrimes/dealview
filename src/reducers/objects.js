// @flow

import _ from 'lodash'
import type { Reducer, State, Action } from 'redux'

import type { ObjectMap } from 'types/commonTypes'
// import type { AssetAction } from 'actions/assets'

export type ObjectStoreState = {
  +status: 'uninitialised'|'loading'|'loaded'|'error',
  +error?: string,
  +objects: ObjectMap
}
export const InitialObjectStoreState: ObjectStoreState = {
  status: 'uninitialised',
  objects: {},
}

export const createObjectsReducer = (name: string, pluralName: string): Reducer<State, Action> => {
  return (state = InitialObjectStoreState, action) => {
    switch (action.type) {
      case `PUT_${name}_REQUEST`:
      case `DELETE_${name}_REQUEST`:
      case `LOAD_${pluralName}_REQUEST`:
        return {
          ...state,
          status: 'loading',
        }
      case `PUT_${name}_SUCCESS`:
        return {
          ...state,
          status: 'loaded',
          objects: { ...state.objects, [action.object.id]: action.object },
        }
      case `DELETE_${name}_SUCCESS`:
        return {
          ...state,
          status: 'loaded',
          objects: _.omit(state.objects, action.id),
        }
      case `LOAD_${pluralName}_SUCCESS`:
        return {
          ...state,
          status: 'loaded',
          objects: action.objects,
        }
      case `PUT_${name}_FAILURE`:
      case `DELETE_${name}_FAILURE`:
      case `LOAD_${pluralName}_FAILURE`:
        return {
          ...state,
          status: 'error',
          error: action.error,
        }
      default:
        return state
    }
  }
}
