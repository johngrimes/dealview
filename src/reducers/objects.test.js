import { createObjectsReducer } from './objects.js'

describe('createObjectsReducer', () => {
  const validObject1 = {
    aspectA: 'foo',
    aspectB: 'bar',
  }
  const validObjectWithId1 = { ...validObject1, id: 'xyz123' }

  const validObject2 = {
    aspectA: 'grate',
    aspectB: 'pad',
  }
  const validObjectWithId2 = { ...validObject2, id: 'abc789' }

  const initialState = {
    status: 'loaded',
    objects: {
      [validObjectWithId1.id]: validObjectWithId1,
    },
  }

  const reducer = createObjectsReducer('RABBIT', 'RABBITS')

  it('should update state upon request', () => {
    const actions = [
      { type: 'PUT_RABBIT_REQUEST', object: validObject2 },
      { type: 'DELETE_RABBIT_REQUEST', id: 'xyz123' },
      { type: 'LOAD_RABBITS_REQUEST' },
    ]
    actions.forEach(action => {
      const nextState = reducer(initialState, action)
      expect(nextState).toMatchSnapshot()
    })
  })

  it('should update state upon put success', () => {
    const action = { type: 'PUT_RABBIT_SUCCESS', object: validObjectWithId2 }
    const nextState = reducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon delete success', () => {
    const action = { type: 'DELETE_RABBIT_SUCCESS', id: 'xyz123' }
    const nextState = reducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon load success', () => {
    const action = {
      type: 'LOAD_RABBITS_SUCCESS',
      objects: {
        [validObjectWithId1.id]: validObjectWithId1,
        [validObjectWithId2.id]: validObjectWithId2,
      },
    }
    const nextState = reducer(initialState, action)
    expect(nextState).toMatchSnapshot()
  })

  it('should update state upon failure', () => {
    const actions = [
      { type: 'PUT_RABBIT_FAILURE', error: 'Something borked' },
      { type: 'DELETE_RABBIT_FAILURE', error: 'Flatlined' },
      { type: 'LOAD_RABBITS_FAILURE', error: 'No rabbits to be found' },
    ]
    actions.forEach(action => {
      const nextState = reducer(initialState, action)
      expect(nextState).toMatchSnapshot()
    })
  })
})
