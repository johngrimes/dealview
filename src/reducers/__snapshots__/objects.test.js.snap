// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createObjectsReducer should update state upon delete success 1`] = `
Object {
  "objects": Object {},
  "status": "loaded",
}
`;

exports[`createObjectsReducer should update state upon failure 1`] = `
Object {
  "error": "Something borked",
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "error",
}
`;

exports[`createObjectsReducer should update state upon failure 2`] = `
Object {
  "error": "Flatlined",
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "error",
}
`;

exports[`createObjectsReducer should update state upon failure 3`] = `
Object {
  "error": "No rabbits to be found",
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "error",
}
`;

exports[`createObjectsReducer should update state upon load success 1`] = `
Object {
  "objects": Object {
    "abc789": Object {
      "aspectA": "grate",
      "aspectB": "pad",
      "id": "abc789",
    },
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "loaded",
}
`;

exports[`createObjectsReducer should update state upon put success 1`] = `
Object {
  "objects": Object {
    "abc789": Object {
      "aspectA": "grate",
      "aspectB": "pad",
      "id": "abc789",
    },
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "loaded",
}
`;

exports[`createObjectsReducer should update state upon request 1`] = `
Object {
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "loading",
}
`;

exports[`createObjectsReducer should update state upon request 2`] = `
Object {
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "loading",
}
`;

exports[`createObjectsReducer should update state upon request 3`] = `
Object {
  "objects": Object {
    "xyz123": Object {
      "aspectA": "foo",
      "aspectB": "bar",
      "id": "xyz123",
    },
  },
  "status": "loading",
}
`;
